eos-payg-csv/eos-payg-csv

```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env python3
+
+import csv
+import os
+import shutil
+import subprocess
+import sys
+
+PERIODS = "1d 2d 3d 4d 5d 30d 60d 90d 365d infinite".split()
+TMPDIR = "tmp"
+OUTDIR = "codes"
+
+def period_to_display_str(period):
+    if period == "infinite":
+        return "infinite"
+
+    days, _ = period.split("d")
+
+    if days == "1":
+        return "1 day"
+
+    return "{0} days".format(days)
+
+def key_to_tmp_filename(key):
+    return os.path.join(TMPDIR, "{}.key".format(key))
+
+def rm_and_mkdir(dirname):
+    if os.path.isdir(dirname) and not os.path.islink(dirname):
+        if len(os.listdir(dirname)) == 0:
+            os.rmdir(dirname)
+        else:
+            shutil.rmtree(dirname)
+    os.mkdir(dirname)
+
+def main(argv):
+    files_written = []
+    id_to_key_file = argv[0]
+
+    id_to_key = {}
+    with open(id_to_key_file, 'r') as f:
+        reader = csv.reader(f, quotechar='"')
+        for row in reader:
+            # we specifically want the last key for each device ID since there
+            # may be multiple entries for each device ID. Using a dict
+            # guarantees this.
+            id_to_key[row[0]] = row[-1]
+
+    # strip out an entry for the CSV header, if it exists
+    id_to_key.pop("device_id", None)
+
+    try:
+        rm_and_mkdir(TMPDIR)
+        rm_and_mkdir(OUTDIR)
+    except OSError:
+        print ("Creation of the directory %s failed" % TMPDIR)
+
+    for id in id_to_key:
+        key_filename = key_to_tmp_filename(id)
+        with open(key_filename, 'w') as f:
+            f.write(id_to_key[id])
+
+        period_to_codes = {}
+        for period in PERIODS:
+            output = subprocess.run(['eos-payg-generate', key_filename, period],
+                    stdout=subprocess.PIPE).stdout.decode('utf-8')
+            codes = []
+            for code in output.split('\n'):
+                codes.append(code)
```

Comment:
Aren’t these three lines entirely equivalent to `codes = output.split('\n')`?

Reply:
Or `codes = output.splitlines()` :)

Reply:
I'll see your `splitlines()` and raise you a list comprehension (for reasons explained in a comment block above this new code):

```
codes = ["'{}".format(code) for code in output.splitlines()]
```

This is a case where I miss Java's overloaded + operator for simple string joins. Maybe Python has a shorter option than `format()`.